### PR TITLE
upgrade contract from contract fix

### PR DESCRIPTION
**Pull request by andrei-marinica**

### CODE REVIEW

File: arwen/contexts/metering.go [hunk 1]
```diff
@@ -213,7 +213,7 @@ func (context *meteringContext) checkGas(vmOutput *vmcommon.VMOutput) error {
 	gasProvided := context.GetGasProvided()
 
 	if totalGas != gasProvided {
-		logOutput.Error("gas usage mismatch", "total gas", totalGas, "gas provided", gasProvided)
+		logOutput.Trace("gas usage mismatch", "total gas", totalGas, "gas provided", gasProvided)
 		return arwen.ErrInputAndOutputGasDoesNotMatch
```
Comment (sasurobert) [inline arwen/contexts/metering.go:217]: This should have stayed as error in order to catch other cases.

**Review by dorin-iancu — DISMISSED** [review]

**Review by catalinnnn — APPROVED** [review]

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -1380,15 +1381,12 @@ func v1_3_upgradeContract(
 		return
 	}
 
-	gasSchedule := metering.GasSchedule()
-	gasToUse = gasSchedule.ElrondAPICost.AsyncCallStep
```
Comment (iulianpascalau) [inline arwen/elrondapi/elrondei.go:1384]: Are we sure we do not want to consume the async call gas?

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -1380,15 +1381,12 @@ func v1_3_upgradeContract(
...
-	metering.UseGas(gasToUse)
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1385]: API call cost should have stayed

File: arwen/elrondapi/elrondei.go (hunk 2 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go:1384]: We are consuming it, but properly. See `runtime.ExecuteAsyncCall()` called below.

File: arwen/elrondapi/elrondei.go [hunk 4]
```diff
@@ -1402,23 +1400,20 @@ func v1_3_upgradeContract(
...
-		Destination: calledSCAddress,
-		Data:        []byte(finalData),
-		GasLimit:    uint64(gasLimit),
-		ValueBytes:  value,
-	})
-
-	// Instruct Wasmer to interrupt the execution of the caller SC.
-	runtime.SetRuntimeBreakpointValue(arwen.BreakpointAsyncCall)
+	runtime.ExecuteAsyncCall(
+		calledSCAddress,
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1413]: Reading it again and I think this should have been handled with ExecuteOnDestContext. Not with asyncCall

**Review by iulianpascalau — APPROVED** [review]

File: arwen/elrondapi/elrondei.go (hunk 4 above)
Comment (andrei-marinica) [reply arwen/elrondapi/elrondei.go:1413]: the contract could in principle be in another shard, if "change owner" was called

File: arwen/elrondapi/elrondei.go (hunk 3 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go:1385]: See `runtime.ExecuteAsyncCall()` below, it consumes the `AsyncCallStep` cost.

File: arwen/elrondapi/elrondei.go (hunk 4 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go:1413]: `ExecuteOnDestContext` is safe as long as we don't migrate accounts between shards.

**Review by sasurobert — APPROVED** [review]

File: arwen/elrondapi/elrondei.go (hunk 4 above)
Comment (andrei-marinica) [reply arwen/elrondapi/elrondei.go:1413]: I mean, the contract could have its ownership transferred to an account from another shard. The new owner will be the one making the update

**Review by sasurobert — CHANGES_REQUESTED** [review]
Too much revert. Please keep error for Arwen coding errors. For user errors keep trace

**Review by sasurobert — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**